<a href="https://colab.research.google.com/github/AlexeyKatko/ColabKatko/blob/main/%D1%82%D0%B5%D0%BE%D1%80%D0%B8%D1%8F_%D0%B2%D0%B5%D1%80%D0%BE%D1%8F%D1%82%D0%BD%D0%BE%D1%81%D1%82%D0%B8_%D0%B4%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D0%B5%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1-2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Урок 1. Расчет вероятности случайных событий

## Задача 1


Из колоды в 52 карты извлекаются случайным образом 4 карты. a) Найти вероятность того, что все карты – крести. б) Найти вероятность, что среди 4-х карт окажется хотя бы один туз.

### Решение 1.1
a) В колоде из 52 листов $\\С\frac{4}{52}=13$ карт одной масти.
Воспользуемся формулой подсчета количества сочетаний по $k$ элементов из множества $n$:
$\\ C^k_n=С\frac{n!}{k!(n-k)!}$

И классической формулой вероятности:
  $\\ P(A)=С\frac{m}{n}$
  
  где $m$-число **благоприятных** исходов, $n$-общее число **всех** исходов.
в нашем случае $\\P (A)=C\frac{4}{52}$-т.е. количество способов извлечь 4 карты из 52, а $\\ m=C\frac{4}{13}$, способов извлечь 4 карты одной масти.
Для облегчения расчетов напишем функцию:

In [2]:
from math import factorial
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

In [3]:
# Вычислим m
m=combinations(13, 4)
print(f'm = {m}')

m = 715


In [4]:
# Вычислим n
n=combinations(52, 4)
print(f'n = {n}')

n = 270725


In [5]:
# Найдем вероятность Р(4крести)
P=m/n
print(f'P(4 крести) = {round(P,4)}')

P(4 крести) = 0.0026


Либо, из классической формулы вероятности следует, что вероятность извлечь

первой картой из 4-х крестовую: 13/52,

вторую крестовую: 12/51;

третью крестовую: 11/52;

четвертую: 10/49.

А поскольку, по условию и первая, и вторая, и третья, и четвертая должны быть крестовые, то P(4крести)
Вычислим:

In [6]:
P1=13/52*12/51*11/50*10/49
print(f'P(4 крести) = {round(P1,4)}')

P(4 крести) = 0.0026


#### Решение 1.2

б) в этом случае так же общее число исходов n=C\frac{4}{52}
, а
 будет складываться из 4 вариантов:

1. из 4-х карт 1 туз:
число способов достать 1 туз из 4-х: С^1_4
, число способов достать три остальные карты С
, общее количество сочетаний 1 туз и 3 не туза
С
из 4-х карт 2 туза:
число способов достать 2 туза из 4-х
, число способов достать две остальные карты С
, общее количество сочетаний 2 туза и 2 не туза
С
из 4-х карт 3 туза:
число способов достать 3 туза из 4-х
, число способов достать оставшуюся карту С
, общее количество сочетаний 3 туза и 1 не туз
С
из 4-х карт 4 туза:
число способов достать 4 туза из 4-х
Общее число благоприятных исходов-сумма этих 4 вариантов.
Вычислим:

In [7]:
# Общее число благоприятных исходов-сумма этих 4 вариантов.
n=combinations(52, 4)
print(f'n = {n}')

n = 270725


In [8]:
m=sum([combinations(4,1)*combinations(48,3),combinations(4,2)*combinations(48,2),combinations(4,3)*combinations(48,1),1])
print(f'm = {m}')

m = 76145


In [9]:
# Найдем вероятность Р(хотя бы 1 туз)
P=m/n
print(f'P(хотя бы 1 туз) = {round(P,4)}')

P(хотя бы 1 туз) = 0.2813


Ответ:

а) Р(4 трефы) ≈ 0.0026 ≈ 0.26%;

б) P(хотя бы 1 туз) ≈ 0.2813 ≈ 28.13%
.

# Задача 2

На входной двери подъезда установлен кодовый замок, содержащий десять кнопок с цифрами от 0 до 9. Код содержит три цифры, которые нужно нажать одновременно. Какова вероятность того, что человек, не знающий код, откроет дверь с первой попытки?

## Решение 2.1

Так как кнопки нажимаются одновременно, то порядок не важен, и для формулы вероятности $\\ P(A)=С\frac{m}{n}$

 общее число исходов будет определяться количиством сочетаний 3-х элементов из 10, $\\ n=С\frac{3}{10}$
, а количество благоприятных исходов m=1
-замок открылся. Вычислим P(A)
:

In [10]:
P=1/combinations(10,3)
print(f'P(открыть с первой попытки) = {round(P,4)}')

P(открыть с первой попытки) = 0.0083


Ответ:

P(открыть с первой попытки) ≈ 0.0083 ≈ 0.83%

# Задача 3

В ящике имеется 15 деталей, из которых 9 окрашены. Рабочий случайным образом извлекает 3 детали. Какова вероятность того, что все извлеченные детали окрашены?

## Решение 3.1

Общее количество исходов будет определяться количеством сочетаний 3 деталей из 15 $\\ n=С\frac{3}{15}$
, а количество благоприятных исходов - количеством сочетаний 3-х окрашенных деталей из 9 $\\ n=С\frac{3}{9}$

In [11]:
# Р(3 из 3 окрашены)
P=combinations(9,3)/combinations(15,3)
print(f'P(3 из 3-х окрашены) = {round(P,4)}')

P(3 из 3-х окрашены) = 0.1846


ИЛИ:

In [12]:
print(f'P(3 из 3-х окрашены) = {round(9/15*8/14*7/13,4)}')

P(3 из 3-х окрашены) = 0.1846


Ответ:

P(3 из 3 окрашены) ≈ 0.1846 ≈ 18.46%

# Задача 4

В лотерее 100 билетов. Из них 2 выигрышных. Какова вероятность того, что 2 приобретенных билета окажутся выигрышными?

## Решение 4.1


  $\\ P(оба\ выигрышных)=\frac{1}{C^1_{100}}=\frac{1}{100!}/{2!(100-2)}= \frac{2!*98!}{100!}$


или

$\\P({оба\ выигрышных})=\frac{1}{100}⋅\frac{1}{99}$

In [13]:
P=1/combinations(100,2)
print(f'P(оба выигрышных = {round(P,4)})')

P(оба выигрышных = 0.0002)


Ответ:

P(оба выигрышных) ≈ 0.0002 ≈ 0.02%

# Урок 2. Дискретные распределения вероятностей

## Задача 1

Вероятность того, что стрелок попадет в мишень, выстрелив один раз, равна 0.8. Стрелок выстрелил 100 раз. Найдите вероятность того, что стрелок попадет в цель ровно 85 раз.

### Решение 1.1

Для решения данной задачи воспользуемся формулой Бернулли: $\\ P_n(X=k)=C^k_np^kq^{n-k}$, где n=100k=85p=0.8q=0.2 тогда:
$\\ P_{100}(X=85)=C^{85}_{100}⋅0.8^{85}⋅0.2^{15}= \frac{100}{85!(100-85)!}⋅0.8^{85}⋅0.2^{15}$

In [14]:
from math import factorial, exp
def bernulli(n, k, p):
    comb=factorial(n)/(factorial(k)*factorial(n-k))
    return comb*(p**k)*(1-p)**(n-k)

In [15]:
print(f'P = {bernulli(100,85,0.8): .4f}')

P =  0.0481


Ответ:

Вероятность того, что стрелок попадет в цель ровно 85 раз: P =  0.0481 = 4.81%

## Задача 2

Вероятность того, что лампочка перегорит в течение первого дня эксплуатации, равна 0.0004. В жилом комплексе после ремонта в один день включили 5000 новых лампочек. Какова вероятность, что ни одна из них не перегорит в первый день? Какова вероятность, что перегорят ровно две?

### Решение 2.1

Поскольку в задаче вероятность наступления события мала, а количество испытаний велико, для решения задачи воспользуемся формулой Пуассона:

$\\ P_m= \frac{λ^m}{m!}·𝑒^{-λ}$, где λ среднее количество наступления события за определенную единицу измерений, и $ λ=p·n$

In [16]:
def poisson_distr(m,p,n):
    lamb=p*n
    return exp(-lamb)*(lamb**m)/factorial(m)

a) По условию задания: p = 0.0004, n = 5000, m = 0

тогда:

In [17]:
print(f'P = {poisson_distr(0,0.0004,5000): .4f}')

P =  0.1353


Ответ:

Вероятность того, что ни одна из ламп не перегорит в первый день P =  0.1353 = 13.53%

#### Решение 2.2

б) По условию задания:
p = 0.0004, n = 5000, m = 2
тогда:

In [18]:
print(f'P = {poisson_distr(2,0.0004,5000): .4f}')

P =  0.2707


Ответ:

Вероятность того, что 2 лампы перегорят в первый день P =  0.2707 = 27.07%

## Задача 3

Монету подбросили 144 раза. Какова вероятность, что орел выпадет ровно 70 раз?

Воспользуемся формулой Бернулли $\\ P_n(X=k)=C^k_np^kq^{n-k}$, в нашем случае: n = 144, k = 70, p = 0.5

Вычислим:

In [19]:
print(f'P = {bernulli(144,70,0.5): .4f}')

P =  0.0628


Ответ:
Вероятность того, что при 144-х кратном подбрасывании монетки орёл выпадет ровно 70 раз P =  0.0628 = 6.28%

## Задача 4

В первом ящике находится 10 мячей, из которых 7 - белые. Во втором ящике - 11 мячей, из которых 9 белых. Из каждого ящика вытаскивают случайным образом по два мяча.
- Какова вероятность того, что все мячи белые?
- Какова вероятность того, что ровно два мяча белые?
- Какова вероятность того, что хотя бы один мяч белый?

Для решения данной задачи подсчитаем общее число элементарных исходов при вытаскивании мячей:
- Для первой корзины это
$\\ n_1 = C^2_{10}$
- Для второй
$\\ n_2 = C^2_{11}$
Напишем функцию, для подсчёта количества комбинаций:

In [20]:
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

In [21]:
n1=combinations(10,2)
n2=combinations(11,2)
print(f'Общее число элементарных исходов для первой корзины: n1 = {n1}\n'
f'Общее число элементарных исходов для второй корзины: n2 = {n2}')

Общее число элементарных исходов для первой корзины: n1 = 45
Общее число элементарных исходов для второй корзины: n2 = 55


### Решение 4.1

a) Для выполнения этого условия необходимо *одновременное* наступление двух событий:
- событие А -из первой корзины извлекли 2 белых шара,
- событие В -из второй корзины извлекли 2 белых шара.

- Для события А число благоприятных элементарных исходов $\\ m_1 = C^2_7$

- Для события B число благоприятных элементарных исходов $\\ m_2 = C^2_9$

Вычислим $m_1$ и $m_2$:

In [22]:
m1=combinations(7,2)
m2=combinations(9,2)
print(f'Число благоприятных элементарных исходов для события А: m1 = {combinations(7,2)}\n'
      f'Число благоприятных элементарных исходов для события В: m2 = {combinations(9,2)}'
     )

Число благоприятных элементарных исходов для события А: m1 = 21
Число благоприятных элементарных исходов для события В: m2 = 36


Для событий А и B вероятности их наступления будут расчитываться как $P(A)=\frac{m_1}{n_1}$ и $P(B)=\frac{m_2}{n_2}$

, а вероятность их одновременного наступления как
$\\ P(A+B)=P(A)·P(B)$
Расчитаем:

In [23]:
print(f'Вероятность наступления события А, Р(А) = {m1/n1: .4f}\n'
      f'Вероятность наступления события В, Р(В) = {m2/n2: .4f}\n'
      f'Вероятность одновременного наступления событий А и В, Р(А+В) = {(m1/n1)*(m2/n2): .4f}'
     )

Вероятность наступления события А, Р(А) =  0.4667
Вероятность наступления события В, Р(В) =  0.6545
Вероятность одновременного наступления событий А и В, Р(А+В) =  0.3055


или решение через условную вероятность
 $P(A+B)=\frac{7}{10}·\frac{6}{9}·\frac{9}{11}·\frac{8}{10}$



In [24]:
P=(7/10)*(6/9)*(9/11)*(8/10)
print(f'Вероятность одновременного наступления событий А и В, Р(А+В) = {P: .4f}')

Вероятность одновременного наступления событий А и В, Р(А+В) =  0.3055


Ответ:
Вероятность того, что все извлечённые мячи белые: P ≈ 0.3055 ≈ 30.55%

#### Решение 4.2

б) Этому варианту соответствует несколько комбинаций:

1. События А -из первой корзины извлекли 2 белых шара и В -из второй корзины извлекли 2 не белых шара, или
2. События С -из первой корзины извлекли 1 белый и 1 не белый шар и D -из второй корзины извлекли 1 белый и 1 не белый шар, или
3. События Е -из первой корзины извлекли 2 не белых шара и F -из второй корзины извлекли 2 белых шара

А вероятность того, что сложится одна из этих комбинаций, будет суммой вероятностей этих комбинаций. Т.е.:
P = P(A)·P(B) + P(C)·P(D) + P(E)·P(F)
Число благоприятных элементарных исходов для каждого события в комбинации:
1. $m_A=C\frac{2}{7}, m_B=C\frac{0}{9}$
2. $m_C=C\frac{1}{7}·C\frac{1}{3}, m_D=C\frac{1}{9}·C\frac{1}{2}$
3. $m_E=C\frac{0}{7}, m_F=C\frac{2}{9}$


⟹ $P=\frac{m_A}{n_1}·\frac{m_B}{n_2}+\frac{m_C}{n_1}·\frac{m_D}{n_2}+\frac{m_E}{n_1}·\frac{m_F}{n_2} = \frac{C^2_7}{n_1}·\frac{C^0_9}{n_2}+\frac{C^1_7·C^1_3}{n_1}·\frac{C^1_9·C^1_2}{n_2}+\frac{C^0_7}{n_1}·\frac{C^2_9}{n_2}$


In [25]:
m_A=combinations(7,2)
m_B=combinations(2,2)
m_C=combinations(7,1)*combinations(3,1)
m_D=combinations(9,1)*combinations(2,1)
m_E=combinations(3,2)
m_F=combinations(9,2)
print(f'Число благоприятных элементарных исходов для события А: m_A = {m_A}\n'
      f'Число благоприятных элементарных исходов для события B: m_B = {m_B}\n'
      f'Число благоприятных элементарных исходов для события C: m_C = {m_C}\n'
      f'Число благоприятных элементарных исходов для события D: m_D = {m_D}\n'
      f'Число благоприятных элементарных исходов для события E: m_E = {m_E}\n'
      f'Число благоприятных элементарных исходов для события F: m_F = {m_F}'
     )

Число благоприятных элементарных исходов для события А: m_A = 21
Число благоприятных элементарных исходов для события B: m_B = 1
Число благоприятных элементарных исходов для события C: m_C = 21
Число благоприятных элементарных исходов для события D: m_D = 18
Число благоприятных элементарных исходов для события E: m_E = 3
Число благоприятных элементарных исходов для события F: m_F = 36


In [26]:
P=(m_A/n1)*(m_B/n2)+(m_C/n1)*(m_D/n2)+(m_E/n1)*(m_F/n2)
print(f'Вероятность того, что будут вытащены ровно 2 белых мяча Р = {P: .4f}')

Вероятность того, что будут вытащены ровно 2 белых мяча Р =  0.2048


или через условную вероятность $P=\frac{7}{10}·\frac{6}{9}·\frac{2}{11}·\frac{1}{10}+4·(\frac{7}{10}·\frac{3}{9}·\frac{9}{11}·\frac{2}{10})+\frac{3}{10}·\frac{2}{9}·\frac{9}{11}·\frac{8}{10}$

где 4 - количество способов достать из каждой корзины 1 белый и 1 не белый мяч.

In [27]:
P1=((7/10)*(6/9)*(2/11)*(1/10))+4*((7/10)*(3/9)*(9/11)*(2/10))+((3/10)*(2/9)*(9/11)*(8/10))
print(f'Вероятность того, что будут вытащены ровно 2 белых мяча Р = {P1: .4f}')

Вероятность того, что будут вытащены ровно 2 белых мяча Р =  0.2048


Ответ:

Вероятность того, что будут вытащены ровно 2 белых мяча: Р ≈ 0.2048 ≈ 20.48%

##### Решение 4.3

в) В данном случае удобнее рассмотреть событие $̄𝔸$
-"не было извлечено ни одного белого мяча", которое наряду с искомым событием А
-"был извлечен хотя бы 1 белый мяч", составляют полную группу событий, по скольку в результате испытаний может произойти только либо событие А
, либо событие $𝔸$
, и поэтому:
$P(A)=1-P(𝔸)$
Для наступления события 𝔸
 необходимо чтобы из первой и второй корзин не было вынуто ни одного белого мяча, число таких благоприятных исходов для первой корзины $m_1=C\frac{2}{3}$
 а для второй
$C\frac{2}{2}$

In [28]:
m1=combinations(3,2)
m2=combinations(2,2)
print(f'Число благоприятных элементарных исходов для первой корзины: m1 = {combinations(3,2)}\n'
      f'Число благоприятных элементарных исходов для второй корзины: m2 = {combinations(2,2)}'
     )

Число благоприятных элементарных исходов для первой корзины: m1 = 3
Число благоприятных элементарных исходов для второй корзины: m2 = 1


Тогда $P(𝔸)=\frac{m_1}{n_1}·\frac{m_2}{n_2}$, вычислим $P(𝔸)$ и $P(A)$

In [29]:
print(f'Вероятность того, что не будет извлечено ни одного белого мяча Р(А_обр) = {(m1/n1)*(m2/n2): .4f}\n'
      f'Вероятность того, что будет извлечен хотя бы один белый мяч Р(А) = 1 - Р(А_обр) = {1-(m1/n1)*(m2/n2): .4f}'
     )

Вероятность того, что не будет извлечено ни одного белого мяча Р(А_обр) =  0.0012
Вероятность того, что будет извлечен хотя бы один белый мяч Р(А) = 1 - Р(А_обр) =  0.9988


Вариант с условной вероятностью: $P(A)=1-\frac{3}{10}·\frac{2}{9}·\frac{2}{11}·\frac{2}{11}·\frac{1}{10}$

In [30]:
P=1-(3/10)*(2/9)*(2/11)*(1/10)
print(f'Вероятность того, что будет извлечен хотя бы один белый мяч Р(А) = {P: .4f}')

Вероятность того, что будет извлечен хотя бы один белый мяч Р(А) =  0.9988


Ответ:

Вероятность того, что будет извлечен хотя бы один белый мяч Р(А) ≈  0.9988 ≈ 99.88%